### PR TITLE
KO-149: Added support for custom kubernetes cluster domain and dnsConfig

### PR DESCRIPTION
**Pull request by abhishekdwivedi3060**

### CODE REVIEW

File: controllers/statefulset.go [hunk 1]
```diff
@@ -1132,6 +1132,9 @@ func (r *SingleClusterReconciler) updateSTSFromPodSpec(
 	st.Spec.Template.ObjectMeta.Annotations = r.aeroCluster.Spec.PodSpec.AerospikeObjectMeta.Annotations
 
 	st.Spec.Template.Spec.DNSPolicy = r.aeroCluster.Spec.PodSpec.DNSPolicy
+	if r.aeroCluster.Spec.PodSpec.DNSConfig != nil {
```
Comment (sud82) [inline controllers/statefulset.go:1135]: Due to this `if check`, the user will not be able to reset the DNS config to nil.

File: test/cluster_test.go [hunk 2]
```diff
@@ -967,6 +991,36 @@ func negativeUpdateClusterValidationTest(
...
+				"InvalidDNSConfiguration", func() {
+					It(
+						"InvalidDnsPolicy: should fail when dnsPolicy is set to 'Default'",
+						func() {
+							aeroCluster, err := getCluster(
+								k8sClient, ctx, clusterNamespacedName,
+							)
+							Expect(err).ToNot(HaveOccurred())
+							aeroCluster.Spec.PodSpec.DNSPolicy = v1.DNSDefault
+							err = deployCluster(k8sClient, ctx, aeroCluster)
```
Comment (sud82) [inline test/cluster_test.go:1004]: Update cluster

File: test/cluster_test.go [hunk 3]
```diff
@@ -967,6 +991,36 @@ func negativeUpdateClusterValidationTest(
...
+
+					It(
+						"MissingDnsConfig: Should fail when dnsPolicy is set to 'None' and no dnsConfig given",
+						func() {
+							aeroCluster, err := getCluster(
+								k8sClient, ctx, clusterNamespacedName,
+							)
+							Expect(err).ToNot(HaveOccurred())
+							aeroCluster.Spec.PodSpec.DNSPolicy = v1.DNSNone
+							err = deployCluster(k8sClient, ctx, aeroCluster)
```
Comment (sud82) [inline test/cluster_test.go:1017]: Update cluster

File: test/cluster_test.go [hunk 4]
```diff
@@ -894,6 +894,30 @@ func negativeDeployClusterValidationTest(
...
+						func() {
+							aeroCluster := createDummyAerospikeCluster(clusterNamespacedName, 1)
+							aeroCluster.Spec.PodSpec.DNSPolicy = v1.DNSDefault
+							err := deployCluster(k8sClient, ctx, aeroCluster)
+							Expect(err).Should(HaveOccurred())
+						},
+					)
+
+					It(
+						"MissingDnsConfig: should fail when dnsPolicy is set to 'None' and no dnsConfig given",
```
Comment (sud82) [inline test/cluster_test.go:911]: Add a positive test case where dnsPolicy is `none` and dnsConfig is given

**Review by sud82 — CHANGES_REQUESTED** [review]

File: controllers/statefulset.go (hunk 1 above)
Comment (tanmayja) [reply controllers/statefulset.go:1135]: can we make this check over DNSPolicy?

File: api/v1beta1/aerospikecluster_validating_webhook.go [hunk 5]
```diff
@@ -1607,3 +1611,15 @@ func ValidateAerospikeObjectMeta(aerospikeObjectMeta *AerospikeObjectMeta) error
 	}
 	return nil
 }
+
+func validateDNS(dnsPolicy v1.DNSPolicy, dnsConfig *v1.PodDNSConfig) error {
+	if dnsPolicy == v1.DNSDefault {
```
[the remark this answers is not in the document]
Comment (abhishekdwivedi3060) [reply api/v1beta1/aerospikecluster_validating_webhook.go:1616]: `Default` policy is not going to work out of the box for us as we are dependent on cluster internal DNS lookup.
It might be possible only if coreDNS related configuration is set across all nodes, which is very unlikely.

[the remark this answers is not in the document]
Comment (abhishekdwivedi3060) [reply api/v1beta1/aerospikecluster_validating_webhook.go:1616]: `Default` policy is not going to work out of the box for us as we are dependent on cluster internal DNS lookup.
It might be possible only if coreDNS related configuration is set across all nodes, which is very unlikely.

File: test/cluster_test.go (hunk 4 above)
Comment (abhishekdwivedi3060) [reply test/cluster_test.go:1000]: Done

File: controllers/statefulset.go (hunk 1 above)
Comment (abhishekdwivedi3060) [reply controllers/statefulset.go]: Removed if condition to handle both create and update scenario

**Review by tanmayja — APPROVED** [review]
